listenbrainz/mbid_mapping_writer/mbid_mapper.py

```diff
@@ -0,0 +1,283 @@
+from listenbrainz.mbid_mapping_writer.stop_words import ENGLISH_STOP_WORDS
+
+DEFAULT_TIMEOUT = 2
+COLLECTION_NAME = "mbid_mapping_latest"
+MATCH_TYPES = ('no_match', 'low_quality', 'med_quality', 'high_quality', 'exact_match')
+MATCH_TYPE_NO_MATCH = 0
+MATCH_TYPE_LOW_QUALITY = 1
+MATCH_TYPE_MED_QUALITY = 2
+MATCH_TYPE_HIGH_QUALITY = 3
+MATCH_TYPE_EXACT_MATCH = 4
+
+MATCH_TYPE_HIGH_QUALITY_MAX_EDIT_DISTANCE = 2
+MATCH_TYPE_MED_QUALITY_MAX_EDIT_DISTANCE = 5
+
+ENGLISH_STOP_WORD_INDEX = {k: 1 for k in ENGLISH_STOP_WORDS}
+
+
+def prepare_query(text):
+    return unidecode(re.sub(" +", " ", re.sub(r'[^\w ]+', '', text)).strip().lower())
+
+
+class MBIDMapper():
+    """
+        This class performs a lookup of one or more artist credit name and recording name pairs
+        to find the best possible match in MusicBrainz. This query will unaccent query
+        parameters, be resistant to typos and attempt to modify the search parameters
+        in effort to find a best match.
+
+        For instance, if "Glory Box feat. Sloppy Jo" isn't matched, it will try to match "Glory Box".
+        It is the caller's responsibility to ensure that the results meet the criteria for the
+        given task. This mapping lookup aims to find a match by a number of means necessary that
+        may or may not be acceptable to the caller.
+
+        Other possible detunings:
+        - Swap artist/recording
+        - remove track numbers from recording
+
+        Overall these detunings and edit distances will need more tuning as we start using this
+        endpoint for our MessyBrainz mapping. For this reason the debug flag and optional
+        debug output remains since more debugging/tuning will need to be done later on.
+    """
+
+    MATCH_TYPE_HIGH_QUALITY_MAX_EDIT_DISTANCE = 2
+    MATCH_TYPE_MED_QUALITY_MAX_EDIT_DISTANCE = 5
+
+    def __init__(self, timeout=DEFAULT_TIMEOUT, remove_stop_words=False, debug=False):
+        self.debug = debug
+        self.log = []
+
+        self.client = typesense.Client({
+            'nodes': [{
+                'host': config.TYPESENSE_HOST,
+                'port': config.TYPESENSE_PORT,
+                'protocol': 'http',
+            }],
+            'api_key': config.TYPESENSE_API_KEY,
+            'connection_timeout_seconds': timeout
+        })
+        self.remove_stop_words = remove_stop_words
+        self.is_ac_detuned = False
+        self.is_r_detuned = False
+
+    def _log(self, str):
+        if self.debug:
+            self.log.append(str)
+
+    def read_log(self):
+        log = self.log
+        self.log = []
+
+        return log
+
+    def detune_query_string(self, query, is_artist_credit):
+        """
+            Detune (remove known extra cruft) a metadata field. If a known
+            trailing string exists in the input string, that string and everything
+            after it is removed.
+        """
+
+        strings = ["(", "[", " ft ", " ft. ", " feat ", " feat. ", " featuring "]
```

Comment:
I would also suggest " - " with spaces around. Would that be too agressive?
Most of the missing matches I see now have some form of " - {year}", " - alternate take", " - Radio Edit" and the like.

The matching has improved in big leap, enough that only those are the only missing matches I see now, making the pattern quite visible. The explainer tool is also super helpful to confirm that intuition.


Compare for example this query with and without the " - 2015 Remaster" at the end:

https://labs.api.listenbrainz.org/explain-mbid-mapping?artist_credit_name=Love&recording_name=Maybe+the+People+Would+Be+the+Times+or+Between+Clark+and+Hilldale+-+2015+Remaster

https://labs.api.listenbrainz.org/explain-mbid-mapping?artist_credit_name=Love&recording_name=Maybe+the+People+Would+Be+the+Times+or+Between+Clark+and+Hilldale

Reply:
Let's give it a try!